src: add `--run-from` runtime flag

mertcanaltin opened the pull request:
add ```--run-from``` runtime flag

Fixes #57489

aduh95 commented:
The commit message doesn't adhere to our guidelines, it should start with an imperative verb so e.g. ``src: add `--run-in` runtime flag``

anonrig commented on src/node_task_runner.cc:
```diff
@@ -206,7 +206,12 @@ void ProcessRunner::OnExit(int64_t exit_status, int term_signal) {
 void ProcessRunner::Run() {
   // keeps the string alive until destructor
-  cwd = package_json_path_.parent_path().string();
+  std::string working_dir;
+  if (!node::per_process::cli_options->run_in.empty())
+    working_dir = node::per_process::cli_options->run_in;
+  else
+    working_dir = package_json_path_.parent_path().string();
+  cwd = working_dir;
```
```suggestion
  if (!node::per_process::cli_options->run_in.empty()) {
    cwd = node::per_process::cli_options->run_in;
  } else {
    cwd = package_json_path_.parent_path().string();
  }
```

anonrig commented on doc/api/cli.md:
```diff
@@ -2298,6 +2298,16 @@ The following environment variables are set when running a script with `--run`:
 
+### `--run-in=<dir>`
+
+Run a package.json script in a custom working directory.
```
```suggestion
Run a package.json script in a specific working directory.
```

anonrig commented on doc/api/cli.md:
```diff
@@ -2298,6 +2298,16 @@ The following environment variables are set when running a script with `--run`:
 * `NODE_RUN_PACKAGE_JSON_PATH`: The path to the `package.json` that is being
   processed.
 
+### `--run-in=<dir>`
```
I'm not sure about this, but:

```suggestion
### `--run-in-directory=<dir>`
```

anonrig replied:
If you accept this change, you need to also change the implementation.

aduh95 replied:
I think I prefer `--run-in`, it's shorter and hopefully should be self explanatory enough when associated to a path

cjihrig replied:
Just throwing these ideas out there (feel free to ignore) based on other existing APIs and CLI flags already in Node: `--run-cwd`, `--run-dir`

mertcanaltin replied:
I updated it to ```--run-dir```

anonrig commented:
We need to also change the pull-request title and description to reflect the changes.

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -222,4 +222,19 @@ describe('node --run [command]', () => {
+    assert.strictEqual(child.stderr, '');
+    assert.strictEqual(child.code, 0);
+  });
+
```
```suggestion
```

mertcanaltin replied:
thanks for support

anonrig commented on src/node_task_runner.cc:
```diff
@@ -206,7 +206,17 @@ void ProcessRunner::OnExit(int64_t exit_status, int term_signal) {
+      init_result->exit_code_ = ExitCode::kGenericUserError;
+      return;
+    }
+  } else {
```
```suggestion
  if (!node::per_process::cli_options->run_in_directory.empty()) {
    cwd = node::per_process::cli_options->run_in_directory;
    if (!std::filesystem::is_directory(cwd)) {
      fprintf(stderr, "Error: %s is not a directory\n", cwd.c_str());
      init_result->exit_code_ = ExitCode::kGenericUserError;
      return;
    }
  } else {
```

anonrig commented on src/node_options.cc:
```diff
@@ -1216,6 +1216,9 @@ PerProcessOptionsParser::PerProcessOptionsParser(
             "Run a script specified in package.json",
             &PerProcessOptions::run);
+  AddOption("--run-in-directory",
+            "Run a package.json script in a custom working directory",
```
```suggestion
            "Run a package.json script in a specific working directory",
```

anonrig reviewed (approved):
LGTM if tests pass (and the new requested test is added)

mertcanaltin commented on test/parallel/test-node-run.js:
```diff
@@ -222,4 +222,18 @@ describe('node --run [command]', () => {
+    const customDir = fixtures.path('run-script');
+    const child = await common.spawnPromisified(
+      process.execPath,
+      [ '--run-in-directory', customDir, '--run', `pwd${envSuffix}` ],
```
I added, thanks

GeoffreyBooth commented:
> so e.g. `` src: add `--run-in` runtime flag ``

I find the name `--run-dir` confusing; you're not running a directory, you're specifying the folder to look in to find the `package.json` file. Why not just ask for the path to the `package.json` file directly? Or allow either a path to the file or the path to the containing folder. Something like:

```shell
node --run-path=./app/package.json --run test:backend
# Or
node --run-path=./app --run test:backend
```

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -223,15 +224,19 @@ describe('node --run [command]', () => {
   });
 
-  it('runs script in a custom working directory using --run-in-directory', async () => {
+  it('runs script in a custom working directory using --run-dir', async () => {
```
You need to add a new test that passes a non-directory path to --run-dir and you should validate the output of it.

mertcanaltin replied:
thanks

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -233,12 +233,25 @@ describe('node --run [command]', () => {
   });
+
+  it('returns error on non-directory path for --run-dir', async () => {
+    const nonDirPath = fixtures.path('run-script/non-directory-file');
+    fs.writeFileSync(nonDirPath, 'dummy content');
```
Just use a package.json file inside fixtures

mertcanaltin replied:
thanks

mertcanaltin commented:
> > so e.g. `` src: add `--run-in` runtime flag ``
> 
> I find the name `--run-dir` confusing; you're not running a directory, you're specifying the folder to look in to find the `package.json` file. Why not just ask for the path to the `package.json` file directly? Or allow either a path to the file or the path to the containing folder. Something like:
> 
> ```shell
> node --run-path=./app/package.json --run test:backend
> # Or
> node --run-path=./app --run test:backend
> ```

I understand your concern, I will wait for a few more ideas to make a collective decision, thank you very much for your suggestion 🙏

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -222,4 +223,35 @@ describe('node --run [command]', () => {
+  });
+
+  it('returns error on non-directory path for --run-dir', async () => {
+    const nonDirPath = fixtures.path('run-script/non-directory-file');
+    fs.writeFileSync(nonDirPath, 'run-script/package.json');
```
```suggestion
    const nonDirPath = fixtures.path('run-script/package.json');
```

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -5,6 +5,7 @@ common.requireNoPackageJSONAbove();
 
 const { it, describe } = require('node:test');
 const assert = require('node:assert');
+const fs = require('node:fs');
```
```suggestion
```

GeoffreyBooth commented on doc/api/cli.md:
```diff
@@ -2294,6 +2294,16 @@ The following environment variables are set when running a script with `--run`:
   processed.
 
+### `--run-dir=<dir>`
+
```
```suggestion

> Stability: 1.0 - Early development

```

And should we add `experimental` to the flag name?

mertcanaltin replied:
I saw that we added to the others

mertcanaltin commented on src/node_options.cc:
```diff
@@ -1216,6 +1216,9 @@ PerProcessOptionsParser::PerProcessOptionsParser(
             "Run a script specified in package.json",
             &PerProcessOptions::run);
+  AddOption("--run-dir",
+            "Run a package.json script in a specific working directory",
```
It sets the Node process's working directory

jasnell replied:
hmm... the naming here was discussed in the linked issue. The name `run-dir` was suggested in the case that the option *only* set the cwd for the run command. If we're going to have this generally set the cwd for the process then it should definitely be named something else

jasnell replied:
That said, the documentation update would imply that it only works with the `--run` command.

targos replied:
If it has a more general effect of setting the initial cwd for the process, and assuming we are ok with such a flag, I think it should be named `--cwd`

GeoffreyBooth replied:
What's the use case for setting the initial cwd for the process? Aside from telling Node where to find `package.json`, which could be done with a more isolated flag.

aduh95 commented on doc/api/cli.md:
```diff
@@ -2294,6 +2294,17 @@ The following environment variables are set when running a script with `--run`:
 
+### `--run-dir=<dir>`
+
+> Stability: 1.0 - Early development
+> Run a package.json script in a specific working directory.
```
```suggestion
<!-- YAML
added: REPLACEME
-->

> Stability: 1.0 - Early development

Run a package.json script in a specific working directory.
```

mertcanaltin replied:
thanks

aduh95 commented on doc/api/cli.md:
````diff
@@ -2294,6 +2294,17 @@ The following environment variables are set when running a script with `--run`:
+
+For example:
+
+```console
+$ node --run-dir=/path/to/dir --run test
````
````suggestion
```bash
node --run-dir=/path/to/dir --run test
````

aduh95 commented:
Can you add a test that uses `--run-in` without `--run` to validate the behavior?

mertcanaltin commented on test/parallel/test-node-run.js:
```diff
@@ -251,12 +251,12 @@ describe('node --run [command]', () => {
   });
 
-  it('returns error when --run-in is used without --run', async () => {
+  it('returns error when --run-dir is used without --run', async () => {
```
I thought this place should use run --dir @aduh95

mertcanaltin replied:
this was for misuse, thank you. I edited it this way:

https://github.com/nodejs/node/pull/57523/commits/bd6f91087936f25ec12ac7185af3d81e73ca3f20

aduh95 commented on test/parallel/test-node-run.js:
```diff
@@ -222,4 +222,46 @@ describe('node --run [command]', () => {
+    assert.strictEqual(child.code, 1);
+  });
+
+  it('returns error when --run-in is used without --run', async () => {
+    const dirPath = fixtures.path('run-script/package.json');
+
+    const child = await common.spawnPromisified(
+      process.execPath,
+      [ '--run-in', dirPath ],
```
```suggestion
  it('returns error when --run-dir is used without --run', async () => {
    const dirPath = fixtures.path('run-script/package.json');

    const child = await common.spawnPromisified(
      process.execPath,
      [ '--run-dir', dirPath ],
```

mertcanaltin replied:
thanks I thought I should test with the wrong command, I thought it was to catch an error

pmarchini commented on src/node_options.cc:
```diff
@@ -1216,6 +1216,9 @@ PerProcessOptionsParser::PerProcessOptionsParser(
   AddOption("--run",
             "Run a script specified in package.json",
             &PerProcessOptions::run);
+  AddOption("--run-dir",
```
I've seen that this has already been raised. While I don't have a better name in mind, I think the current name is misleading

mertcanaltin replied:
I agree with @GeoffreyBooth  suggestion 

```--run-path`` +1

mertcanaltin replied:
I will wait for more advice to be given, please feel free to give feedback 🚀

vmoroz replied:
@mertcanaltin , IMHO the most common term used in such context is `working_directory` or `working_dir`.
E.g. "cwd" stands for "current working directory", "pwd" - "print working directory", etc.

Another example is the `WORKING_DIRECTORY` attribute used by [CMake](https://cmake.org/cmake/help/latest/command/add_custom_target.html). 

The new code would read: `cwd = node::per_process::cli_options->working_dir;`
I believe it is more natural than `cwd = node::per_process::cli_options->run_dir;`

vmoroz replied:
The `--cwd` is also a great name proposed above by @targos

vmoroz replied:
Please disregard my previous posts. I read through the discussion in the linked issue and see why the "working directory" term is not used. I am actually puzzled why the new flag is needed at all. Users can always provide the full path for the `--run` parameter.

mertcanaltin replied:
@vmoroz, thank you very much for your suggestions. I think adding this flag gives users an easy and clear way to specify the correct directory and ensures that commands work in the intended context. This extra flexibility can provide a good experience in scenarios where precise control over the working directory is needed, such as when running new API documentation tools.

aduh95 replied:
May I suggest `--run-from`

ovflowd replied:
As far as I know this argument is supposed to indicate where the package.json is located at, such as a CWD, it is supposed to be like the CLI flag that exists within npm. But maybe Im mistaken here? 🤔

anonrig commented on test/parallel/test-node-run.js:
```diff
@@ -222,4 +222,46 @@ describe('node --run [command]', () => {
+    assert.strictEqual(child.code, 1);
+  });
+
+  it('returns error when --run-dir is used without --run', async () => {
```
on line 263 you validate that stderr is "" and there is no error thrown. How are we testing returning error here?

aduh95 replied:
```suggestion
  it('--run-dir should be no-op when used without --run', async () => {
```

mertcanaltin replied:
I had set this up to use the wrong command, thank you for the editing suggestion

GeoffreyBooth commented:
I'm not opposed to this feature, but there are multiple requests for renaming, from different people: https://github.com/nodejs/node/pull/57523#discussion_r2001144695, https://github.com/nodejs/node/pull/57523#discussion_r2001135291, https://github.com/nodejs/node/pull/57523#discussion_r2001285132, https://github.com/nodejs/node/pull/57523#issuecomment-2731009463. There also seems to be some confusion/lack of consensus about whether this should only affect `--run` or the entire process.

I'd like to see this land in some form, so I don't want to block this, but I think the above should be resolved before someone lands this.

anonrig commented:
> I'd like to see this land in some form, so I don't want to block this, but I think the above should be resolved before someone lands this.

We can land this as experimental, and later iterate on this. New documentation infra changes require this to be implemented, so I recommend merging and iterating.

GeoffreyBooth commented:
>  I recommend merging and iterating.

This flag currently lacks `experimental` in the name. It will be disruptive to users to land one name and then rename it. It won't take long to reach a consensus here; I think it's worth taking the time to do so to avoid the user disruption.

ovflowd reviewed (dismissed):
LGTM

anonrig commented:
> This flag currently lacks experimental in the name. It will be disruptive to users to land one name and then rename it. It won't take long to reach a consensus here; I think it's worth taking the time to do so to avoid the user disruption.

We also landed --run without an experimental flag but made it experimental in the docs. I think we can follow the same path here.

GeoffreyBooth commented:
Removing `author ready` because there are outstanding review comments. I think first we need to decide if we want this flag to apply to where `--run` finds its `package.json`, or to apply to the working directory of the Node process. I and some others assumed the former when I read the PR and its documentation, so if the intent is the latter, I think that needs to be made clearer in the docs. Perhaps the flag shouldn’t have `run` in the name.

Once we answer that question, then we can pick a name. I think a GitHub emoji vote should suffice unless someone feels really strongly. But before we get to the name we need to settle on the purpose.

ovflowd commented:
I'm removing the `author ready` label because there are outstanding review comments. We need to determine whether this flag should apply to the location where `--run` finds its `package.json`, or to the working directory of the Node process. I, along with some others, initially assumed it applied to the former based on the pull request and its documentation.

I agree with this approach; I originally proposed it. The focus is not on the current working directory (CWD) of the process that runs, but rather on the directory where the `package.json` is located from which the scripts should be executed. It should still run from the context you provide. For example:

- If `pwd` returns `/home/cwunder/hello`
- And `package.json` is located at `/home/cwunder/something/package.json`
- Then passing `--run-from /home/cwunder/something` will simply read the `package.json` from that location.
- However, if you execute `node --run my-script --run-from /home/cwunder/something`, `process.cwd()` should still return `/home/cwunder/hello` if invoked within the script.

jasnell commented:
Based on the discussion that prompted opening this PR in the first place, the ask is for a way of setting the working directory of the `--run` command, not the process as a whole.  I've seen nothing to suggest that a command-line flag to set the working directory of the process is necessary. I believe this should be limited to `--run` usage and named accordingly.

ovflowd commented:
> Based on the discussion that prompted opening this PR in the first place, the ask is for a way of setting the working directory of the `--run` command, not the process as a whole. I've seen nothing to suggest that a command-line flag to set the working directory of the process is necessary. I believe this should be limited to `--run` usage and named accordingly.

Yup, can confirm, that was the ask :)

mertcanaltin commented:
thanks a lot, now I am merging all commits and fixing my first commit message 🚀

GeoffreyBooth commented:
> the ask is for a way of setting the working directory of the --run command, not the process as a whole.

What does it mean to apply only to `--run`? Does that just tell it where to find `package.json` or are there other effects?

mertcanaltin commented:
> Reference in new issue
> Report co



> > the ask is for a way of setting the working directory of the --run command, not the process as a whole.
> 
> What does it mean to apply only to `--run`? Does that just tell it where to find `package.json` or are there other effects?

only determines the directory where the --run command finds the package.json file; it does not change the process-wide working directory.

GeoffreyBooth commented:
> only determines the directory where the --run command finds the package.json file; it does not change the process-wide working directory.

Okay, can it accept as input either the path to the containing folder _or_ the path directly to the `package.json` file?

targos commented:
I'm not sure this feature will work as expected if it doesn't set the process's cwd. Most scripts I write wouldn't work FWIW

mertcanaltin commented:
> > only determines the directory where the --run command finds the package.json file; it does not change the process-wide working directory.
> 
> Okay, can it accept as input either the path to the containing folder _or_ the path directly to the `package.json` file?
> 
> (As in, if this isn't the current behavior is it possible to update the PR so that this is possible.)

yes, I'm trying now

mertcanaltin commented:
hello i sent an improvement @GeoffreyBooth @ovflowd @targos 

I tried to implement that if given a package.json directly, this directory is now selected as the parent directory

GeoffreyBooth commented on doc/api/cli.md:
````diff
@@ -2294,6 +2294,22 @@ The following environment variables are set when running a script with `--run`:
 * `NODE_RUN_PACKAGE_JSON_PATH`: The path to the `package.json` that is being
   processed.
 
+### `--run-from=<dir>`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.0 - Early development
+
+Run a script defined in the package.json located in the specified directory.
+
+For example:
+
+```bash
+node --run-from=/path/to/dir --run test
+```
````
```suggestion
### `--run-from=<path>`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.0 - Early development

Run a `package.json` script from a specified path to a `package.json` file
or path to the containing folder of a `package.json` file.

```bash
node --run-from=/app/package.json --run test
# Or
node --run-from=/app/ --run test
```
```

mertcanaltin replied:
thanks

GeoffreyBooth commented on doc/api/cli.md:
```diff
@@ -2294,6 +2294,22 @@ The following environment variables are set when running a script with `--run`:
+
+Run a `package.json` script from a specified path to a `package.json` file
+or path to the containing folder of a `package.json` file.
+
```
Sorry I think you accepted the suggestion while I was still editing.

```suggestion

The script is run from the current working directory, not from the
`package.json` containing folder.

```

GeoffreyBooth commented on doc/api/cli.md:
````diff
@@ -2294,6 +2294,22 @@ The following environment variables are set when running a script with `--run`:
+```bash
+node --run-from=/app/package.json --run test
+# Or
+node --run-from=/app/ --run test
````
I can't add this as a suggestion, but this needs a closing triple backtick.

GeoffreyBooth commented:
> I'm not sure this feature will work as expected if it doesn't set the process's cwd. Most scripts I write wouldn't work FWIW

I'm wondering how this can work in practice too. Does it solve https://github.com/nodejs/node/issues/57489? Are the scripts in that `package.json` runnable from a different CWD?

anonrig commented on src/node_task_runner.cc:
```diff
@@ -206,7 +206,35 @@ void ProcessRunner::OnExit(int64_t exit_status, int term_signal) {
 
 void ProcessRunner::Run() {
   // keeps the string alive until destructor
-  cwd = package_json_path_.parent_path().string();
```
This will not work since this doesn't add the current node_modules to path environment variable.

anonrig commented:
> only determines the directory where the --run command finds the package.json file; it does not change the process-wide working directory.

@mertcanaltin this is wrong. you're also changing the cwd of the process that we are spawning. 

To answer other reviewers: I think this solution has lots of flaws. It doesn't properly propagate path variable (according to the given --run-from), and makes lots of unnecessary system calls.

mertcanaltin commented:
> > only determines the directory where the --run command finds the package.json file; it does not change the process-wide working directory.
> 
> @mertcanaltin this is wrong. you're also changing the cwd of the process that we are spawning.
> 
> To answer other reviewers: I think this solution has lots of flaws. It doesn't properly propagate path variable (according to the given --run-from), and makes lots of unnecessary system calls.

now only the cwd of the spawned process (options_.cwd) is set, without changing the process-wide cwd. Also, if the file path is given for --run-from, the parent directory is used. I hope this addresses your concerns.

mertcanaltin commented on src/node_task_runner.cc:
```diff
@@ -221,7 +221,12 @@ void ProcessRunner::Run() {
       }
     }
-
+    // Adding node_modules/.bin under cwd to PATH environment variable
+    std::filesystem::path nmBin =
+        std::filesystem::path(cwd) / "node_modules" / ".bin";
+    if (std::filesystem::is_directory(nmBin)) {
+      path_env_var_ = nmBin.string() + env_var_separator + path_env_var_;
+    }
```
@anonrig I tried to process the path given with --run-from and add the ```node_modules/.bin``` directory under the cwd to the ```PATH``` Without changing the cwd of the parent process, only the cwd of the spawned process will be set, is this a good solution ?

mertcanaltin commented:
Thank you very much for your valuable contributions and feedback on this PR. Since this work has progressed in @flakey5 's PR, I'm closing this one 🚀
